snakefmt/formatter.py

```diff
@@ -143,9 +148,29 @@ def run_black_format_str(
         try:
             fmted = black.format_str(string, mode=black_mode)
         except black.InvalidInput as e:
-            raise InvalidPython(
-                f"Got error:\n```\n{str(e)}\n```\n" f"while formatting code with black."
-            ) from None
+            err_msg = ""
+            # Not clear whether all Black errors start with 'Cannot parse' - it seems to
+            # in the tests I ran
+            match = re.search(r"(Cannot parse: )(?P<line>\d+)(.*)", str(e))
+            try:
+                next_token = next(self.snakefile)
+                self.snakefile.denext(next_token)
+            except StopIteration:
+                next_token = None
+            if match and next_token is not None:
+                # this is the line number within the piece of code that was passed to
+                # black, not necessarily the line number within the Snakefile
+                line_num = int(match.group("line"))
+                context_line_num = next_token.start[0] - len(string.splitlines())
+                total_line_num = context_line_num + line_num - 1
+                err_msg = match.group(1) + str(total_line_num) + match.group(3)
+            else:
+                err_msg = str(e) + (
+                    "\n\n(Note reported line number may be incorrect, as"
+                    " snakefmt could not determine the true line number)"
+                )
+            err_msg = f"Black error:\n```\n{str(err_msg)}\n```\n"
+            raise InvalidPython(err_msg) from None
```

Comment:
I like what you did and took the liberty of modifying it via a pushed commit to the PR:
* The way you calculated line number gave the line number in the formatted file, which isn't line number in the input file. The new way I do it computes in input file. Note it uses a token lookahead method from the `Snakefile` class (`denext()`) which is quite useful, see comments below for how we may want to use it again in this PR
* I modified black's own error output to change just the line number, I find this more consistent and readable.

Feel free to change again for e.g. the formatting of the error messages